// react-native-select-multiple/src/SelectMultiple.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, ListView, Text, TouchableWithoutFeedback, Image } from 'react-native'
import styles from './SelectMultiple.styles'
import checkbox from '../images/icon-checkbox.png'
import checkboxChecked from '../images/icon-checkbox-checked.png'
import { mergeStyles } from './style'

const itemType = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({ name: PropTypes.any, id: PropTypes.any })
])

const styleType = PropTypes.oneOfType([
  PropTypes.object,
  PropTypes.number,
  PropTypes.array
])

const sourceType = PropTypes.oneOfType([PropTypes.object, PropTypes.number])

// A customiseable ListView that allows you to select multiple rows
export default class SelectMultiple extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(itemType).isRequired,
    selectedItems: PropTypes.arrayOf(itemType),

    onSelectionsChange: PropTypes.func.isRequired,

    checkboxSource: sourceType,
    selectedCheckboxSource: sourceType,
    rendername: PropTypes.func,
    listViewProps: PropTypes.any,
    style: styleType,
    rowStyle: styleType,
    checkboxStyle: styleType,
    nameStyle: styleType,

    selectedRowStyle: styleType,
    selectedCheckboxStyle: styleType,
    selectednameStyle: styleType
  }

  static defaultProps = {
    selectedItems: [],
    style: {},
    rowStyle: {},
    checkboxStyle: {},
    checkboxCheckedStyle: {},
    nameStyle: {},
    checkboxSource: checkbox,
    selectedCheckboxSource: checkboxChecked,
    rendername: null
  }

  constructor (props) {
    super(props)

    const rows = this.getRowData(props)

    const dataSource = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1.id !== r2.id || r1.selected !== r2.selected
    }).cloneWithRows(rows)

    this.state = { dataSource }
  }

  componentWillReceiveProps (nextProps) {
    const rows = this.getRowData(nextProps)
    const dataSource = this.state.dataSource.cloneWithRows(rows)
    this.setState({ dataSource })
  }

  getRowData ({ items, selectedItems }) {
    items = items.map(this.tonameidObject)
    selectedItems = (selectedItems || []).map(this.tonameidObject)

    items.forEach((item) => {
      item.selected = selectedItems.some((i) => i.id === item.id)
    })

    return items
  }

  onRowPress (row) {
    const { name, id } = row
    let { selectedItems } = this.props

    selectedItems = (selectedItems || []).map(this.tonameidObject)

    const index = selectedItems.findIndex((selectedItem) => selectedItem.id === id)

    if (index > -1) {
      selectedItems = selectedItems.filter((selectedItem) => selectedItem.id !== id)
    } else {
      selectedItems = selectedItems.concat({ name, id })
    }

    this.props.onSelectionsChange(selectedItems, { name, id })
  }

  tonameidObject (obj) {
    if (Object.prototype.toString.call(obj) === '[object String]') {
      return { name: obj, id: obj }
    } else {
      return { name: obj.name, id: obj.id }
    }
  }

  render () {
    const { dataSource } = this.state
    const { style, listViewProps } = this.props
    const { renderItemRow } = this
    return <ListView style={style} dataSource={dataSource} renderRow={renderItemRow} {...(listViewProps || {})} />
  }

  rendername = (name, style, selected) => {
    if (this.props.rendername) {
      return this.props.rendername(name, style, selected)
    }
    return (
      <Text style={style}>{name}</Text>
    )
  }

  renderItemRow = (row) => {
    let {
      checkboxSource,
      rowStyle,
      nameStyle,
      checkboxStyle
    } = this.props

    const {
      selectedCheckboxSource,
      selectedRowStyle,
      selectedCheckboxStyle,
      selectednameStyle
    } = this.props

    if (row.selected) {
      checkboxSource = selectedCheckboxSource
      rowStyle = mergeStyles(styles.row, rowStyle, selectedRowStyle)
      checkboxStyle = mergeStyles(styles.checkbox, checkboxStyle, selectedCheckboxStyle)
      nameStyle = mergeStyles(styles.name, nameStyle, selectednameStyle)
    } else {
      rowStyle = mergeStyles(styles.row, rowStyle)
      checkboxStyle = mergeStyles(styles.checkbox, checkboxStyle)
      nameStyle = mergeStyles(styles.name, nameStyle)
    }

    return (
      <TouchableWithoutFeedback onPress={() => this.onRowPress(row)}>
        <View style={rowStyle}>
          <Image style={checkboxStyle} source={checkboxSource} />
          {this.rendername(row.name, nameStyle, row.selected)}
        </View>
      </TouchableWithoutFeedback>
    )
  }
}
